File: frappe-bench/apps/frappe_pim/frappe_pim/pim/api/onboarding.py
"""PIM Onboarding API Endpoints

This module provides API endpoints for the SaaS onboarding wizard.
All functions support both synchronous use and whitelisted API access.

Endpoints:
- start_onboarding: Start/resume onboarding for the current user
- get_onboarding_state: Get current onboarding state and progress
- save_step_data: Save form data for a specific step (partial save)
- apply_archetype_template: Apply an industry archetype template
- complete_onboarding: Advance through steps or complete onboarding
- get_available_archetypes: List available industry archetype templates
- skip_onboarding: Skip the onboarding wizard entirely
- reset_onboarding: Reset onboarding to initial state
- preview_archetype: Preview what an archetype template will create

Note: frappe imports are deferred to function level to allow module
import without Frappe being available (e.g., for testing/verification).
"""


def start_onboarding(user=None):
    """Start or resume onboarding for a user.

    Creates a new PIM Onboarding State document if one doesn't exist
    for the specified user, or returns the existing state.

    Args:
        user: User email. Defaults to current session user.

    Returns:
        dict: Onboarding state summary
            - user: User email
            - current_step: Current step in the wizard
            - is_completed: Whether onboarding is finished
            - is_skipped: Whether onboarding was skipped
            - progress_percent: Completion percentage (0-100)
            - selected_archetype: Chosen industry archetype (if any)
            - template_applied: Whether a template has been applied
            - started_at: When onboarding began
            - completed_at: When onboarding finished
            - completed_steps: List of completed step names
            - next_step: Next step to complete
            - previous_step: Previous step (for back navigation)
            - total_steps: Total number of steps
            - steps: Detailed step list with status

    Example:
        >>> state = start_onboarding()
        >>> print(f"Current step: {state['current_step']}")
    """
    import frappe
    from frappe import _

    if not user:
        user = frappe.session.user

    # Permission check
    if not frappe.has_permission("PIM Onboarding State", "read"):
        frappe.throw(_("Not permitted to access onboarding"), frappe.PermissionError)

    # Find existing state for this user
    existing = frappe.db.exists("PIM Onboarding State", {"user": user})

    if existing:
        doc = frappe.get_doc("PIM Onboarding State", existing)
    else:
        # Create a new onboarding state
        doc = frappe.new_doc("PIM Onboarding State")
        doc.user = user
        doc.current_step = "pending"
        doc.insert(ignore_permissions=True)
        frappe.db.commit()

    # If still at pending, advance to company_info to start
    if doc.current_step == "pending":
        doc.advance_step()

    return doc.get_status_summary()


def get_onboarding_state(user=None):
    """Get the current onboarding state for a user.

    Returns the full onboarding state including progress, step data,
    and template application status.

    Args:
        user: User email. Defaults to current session user.

    Returns:
        dict: Onboarding state summary with all step details.
            Returns a 'not_started' state if no onboarding exists.

    Example:
        >>> state = get_onboarding_state()
        >>> if state['is_completed']:
        ...     print("Onboarding complete!")
        >>> else:
        ...     print(f"Progress: {state['progress_percent']}%")
    """
    import frappe
    from frappe import _

    if not user:
        user = frappe.session.user

    if not frappe.has_permission("PIM Onboarding State", "read"):
        frappe.throw(_("Not permitted to access onboarding"), frappe.PermissionError)

    existing = frappe.db.exists("PIM Onboarding State", {"user": user})

    if not existing:
        return {
            "user": user,
            "current_step": "not_started",
            "is_completed": False,
            "is_skipped": False,
            "progress_percent": 0,
            "selected_archetype": None,
            "template_applied": False,
            "started_at": None,
            "completed_at": None,
            "completed_steps": [],
            "next_step": None,
            "previous_step": None,
            "total_steps": 12,
            "steps": [],
        }

    doc = frappe.get_doc("PIM Onboarding State", existing)
    summary = doc.get_status_summary()

    # Include step data for the current step (useful for resuming forms)
    current_data = doc.get_step_data(doc.current_step)
    if current_data:
        summary["current_step_data"] = current_data

    return summary


def save_step_data(step, form_data, user=None, advance=False):
    """Save form data for a specific onboarding step.

    Allows partial saves (drafts) during a step. Optionally
    advances to the next step after saving.

    Args:
        step: The step name to save data for (e.g., "company_info",
            "industry_selection", "product_structure")
        form_data: JSON string or dict of form data collected during the step
        user: User email. Defaults to current session user.
        advance: If True, also advance to the next step after saving

    Returns:
        dict: Updated onboarding state summary

    Example:
        >>> result = save_step_data(
        ...     step="company_info",
        ...     form_data='{"company_name": "Acme Corp", "industry": "retail"}',
        ...     advance=True
        ... )
        >>> print(f"Now at step: {result['current_step']}")
    """
    import frappe
    from frappe import _
    import json

    if not user:
        user = frappe.session.user

    if not frappe.has_permission("PIM Onboarding State", "write"):
        frappe.throw(_("Not permitted to modify onboarding"), frappe.PermissionError)

    # Find onboarding state
    existing = frappe.db.exists("PIM Onboarding State", {"user": user})
    if not existing:
        frappe.throw(
            _("No onboarding state found for user {0}. Call start_onboarding first.").format(user),
            title=_("Not Found")
        )

    doc = frappe.get_doc("PIM Onboarding State", existing)

    # Parse form_data if it's a JSON string
    if isinstance(form_data, str):
        try:
            parsed_data = json.loads(form_data)
        except (json.JSONDecodeError, TypeError):
            frappe.throw(
                _("Invalid form_data: must be valid JSON"),
                title=_("Invalid Data")
            )
    else:
        parsed_data = form_data

    if not isinstance(parsed_data, dict):
        frappe.throw(
            _("form_data must be a JSON object (dict)"),
            title=_("Invalid Data")
        )

    # Save step data
    doc.save_step_data(step, parsed_data)

    # Optionally advance to next step
    if advance:
        doc.advance_step()

    return doc.get_status_summary()


def apply_archetype_template(archetype, user=None, dry_run=False):
    """Apply an industry archetype template to configure PIM.

    Loads the specified archetype template (and its base template if
    declared) and creates all configuration entities: attribute groups,
    attribute types, attributes, product types, product families, and
    categories.

    Args:
        archetype: Archetype identifier (e.g., "fashion", "industrial",
            "food", "base")
        user: User email. Defaults to current session user.
        dry_run: If True, validate and preview without creating records

    Returns:
        dict: Template application result
            - success: Whether application succeeded
            - archetype: The archetype that was applied
            - status: Application status (completed, partial, failed)
            - entities_created: Number of entities created
            - entities_skipped: Number of entities skipped (already exist)
            - entities_failed: Number of entities that failed
            - details: Per-section breakdown
            - errors: List of error messages
            - messages: List of informational messages

    Example:
        >>> result = apply_archetype_template("fashion")
        >>> if result['success']:
        ...     print(f"Created {result['entities_created']} entities")
    """
    import frappe
    from frappe import _

    if not user:
        user = frappe.session.user

    if not frappe.has_permission("PIM Onboarding State", "write"):
        frappe.throw(_("Not permitted to apply templates"), frappe.PermissionError)

    # Resolve onboarding state name for integration
    onboarding_state_name = None
    existing = frappe.db.exists("PIM Onboarding State", {"user": user})
    if existing:
        onboarding_state_name = existing

    # Apply via the template engine
    from frappe_pim.pim.services.template_engine import TemplateEngine

    try:
        template_result = TemplateEngine.apply_template(
            archetype_name=archetype,
            onboarding_state_name=onboarding_state_name if not dry_run else None,
            dry_run=dry_run,
        )
    except Exception as e:
        frappe.log_error(
            message=f"Template application failed for archetype '{archetype}': {str(e)}",
            title="PIM Onboarding Template Error"
        )
        return {
            "success": False,
            "archetype": archetype,
            "status": "failed",
            "entities_created": 0,
            "entities_skipped": 0,
            "entities_failed": 0,
            "details": {},
            "errors": [str(e)],
            "messages": [],
        }

    result_dict = template_result.to_dict()
    success = result_dict["status"] in ("completed", "partial")

    return {
        "success": success,
        "archetype": result_dict["archetype"],
        "status": result_dict["status"],
        "entities_created": result_dict["entities_created"],
        "entities_skipped": result_dict["entities_skipped"],
        "entities_failed": result_dict["entities_failed"],
        "details": result_dict["details"],
        "errors": result_dict["errors"],
        "messages": result_dict["messages"],
        "dry_run": dry_run,
    }


def complete_onboarding(user=None, form_data=None):
    """Advance the onboarding wizard to the next step, or complete it.

    If on the last actionable step, marks onboarding as completed.
    Saves form data for the current step if provided.

    Args:
        user: User email. Defaults to current session user.
        form_data: Optional JSON string or dict of form data for the
            current step before advancing.

    Returns:
        dict: Updated onboarding state summary

    Example:
        >>> result = complete_onboarding(
        ...     form_data='{"confirm": true}'
        ... )
        >>> if result['is_completed']:
        ...     print("Onboarding finished!")
    """
    import frappe
    from frappe import _
    import json

    if not user:
        user = frappe.session.user

    if not frappe.has_permission("PIM Onboarding State", "write"):
        frappe.throw(_("Not permitted to modify onboarding"), frappe.PermissionError)

    existing = frappe.db.exists("PIM Onboarding State", {"user": user})
    if not existing:
        frappe.throw(
            _("No onboarding state found for user {0}. Call start_onboarding first.").format(user),
            title=_("Not Found")
        )

    doc = frappe.get_doc("PIM Onboarding State", existing)

    if doc.is_completed:
        return doc.get_status_summary()

    # Parse form_data
    parsed_data = None
    if form_data:
        if isinstance(form_data, str):
            try:
                parsed_data = json.loads(form_data)
            except (json.JSONDecodeError, TypeError):
                frappe.throw(
                    _("Invalid form_data: must be valid JSON"),
                    title=_("Invalid Data")
                )
        else:
            parsed_data = form_data

    # Advance to next step (handles completion automatically)
    doc.advance_step(form_data=parsed_data)

    return doc.get_status_summary()


def get_available_archetypes():
    """Get list of available industry archetype templates.

    Scans the PIM fixtures directory for archetype template JSON files
    and returns metadata for each.

    Returns:
        dict: Available archetypes information
            - archetypes: List of archetype dicts, each containing:
                - archetype: Identifier (e.g., "fashion")
                - label: Human-readable name (e.g., "Fashion & Apparel")
                - description: Brief description
                - version: Template version
            - total: Total number of available archetypes

    Example:
        >>> result = get_available_archetypes()
        >>> for arch in result['archetypes']:
        ...     print(f"{arch['label']}: {arch['description']}")
    """
    import frappe
    from frappe import _

    if not frappe.has_permission("PIM Onboarding State", "read"):
        frappe.throw(_("Not permitted to view archetypes"), frappe.PermissionError)

    from frappe_pim.pim.services.template_engine import TemplateEngine

    archetypes = TemplateEngine.get_available_archetypes()

    # Remove internal file_path from response
    for arch in archetypes:
        arch.pop("file_path", None)

    return {
        "archetypes": archetypes,
        "total": len(archetypes),
    }


def skip_onboarding(user=None):
    """Skip the onboarding wizard entirely.

    Marks the onboarding as skipped so the user can configure
    PIM manually without the guided wizard.

    Args:
        user: User email. Defaults to current session user.

    Returns:
        dict: Updated onboarding state summary with is_skipped=True

    Example:
        >>> result = skip_onboarding()
        >>> assert result['is_skipped'] is True
    """
    import frappe
    from frappe import _

    if not user:
        user = frappe.session.user

    if not frappe.has_permission("PIM Onboarding State", "write"):
        frappe.throw(_("Not permitted to modify onboarding"), frappe.PermissionError)

    existing = frappe.db.exists("PIM Onboarding State", {"user": user})
    if not existing:
        frappe.throw(
            _("No onboarding state found for user {0}. Call start_onboarding first.").format(user),
            title=_("Not Found")
        )

    doc = frappe.get_doc("PIM Onboarding State", existing)
    doc.skip_onboarding()

    return doc.get_status_summary()


def reset_onboarding(user=None):
    """Reset onboarding to initial state.

    Clears all step data, template application, and resets progress
    to zero. Requires System Manager role.

    Args:
        user: User email. Defaults to current session user.

    Returns:
        dict: Reset onboarding state summary

    Example:
        >>> result = reset_onboarding()
        >>> assert result['current_step'] == 'pending'
        >>> assert result['progress_percent'] == 0
    """
    import frappe
    from frappe import _

    if not user:
        user = frappe.session.user

    if not frappe.has_permission("PIM Onboarding State", "write"):
        frappe.throw(_("Not permitted to reset onboarding"), frappe.PermissionError)

    existing = frappe.db.exists("PIM Onboarding State", {"user": user})
    if not existing:
        frappe.throw(
            _("No onboarding state found for user {0}").format(user),
            title=_("Not Found")
        )

    doc = frappe.get_doc("PIM Onboarding State", existing)
    doc.reset_onboarding()

    return doc.get_status_summary()


def preview_archetype(archetype):
    """Preview what an archetype template will create without applying it.

    Returns a summary of all entities that would be created, including
    counts and item keys per section.

    Args:
        archetype: Archetype identifier (e.g., "fashion", "industrial")

    Returns:
        dict: Preview information
            - archetype: Archetype identifier
            - label: Human-readable name
            - description: Brief description
            - version: Template version
            - extends: Base template name (if any)
            - sections: Per-section details with counts and item keys

    Example:
        >>> preview = preview_archetype("fashion")
        >>> print(f"Will create {preview['sections']['attributes']['count']} attributes")
    """
    import frappe
    from frappe import _

    if not frappe.has_permission("PIM Onboarding State", "read"):
        frappe.throw(_("Not permitted to preview archetypes"), frappe.PermissionError)

    from frappe_pim.pim.services.template_engine import TemplateEngine

    try:
        return TemplateEngine.preview_template(archetype)
    except FileNotFoundError:
        frappe.throw(
            _("Archetype template '{0}' not found").format(archetype),
            title=_("Template Not Found")
        )
    except ValueError as e:
        frappe.throw(
            _("Invalid template: {0}").format(str(e)),
            title=_("Template Error")
        )


# ============================================================================
# Whitelist Wrapper
# ============================================================================

def _wrap_for_whitelist():
    """Add @frappe.whitelist() decorators at runtime."""
    import frappe

    global start_onboarding, get_onboarding_state, save_step_data
    global apply_archetype_template, complete_onboarding, get_available_archetypes
    global skip_onboarding, reset_onboarding, preview_archetype

    start_onboarding = frappe.whitelist()(start_onboarding)
    get_onboarding_state = frappe.whitelist()(get_onboarding_state)
    save_step_data = frappe.whitelist()(save_step_data)
    apply_archetype_template = frappe.whitelist()(apply_archetype_template)
    complete_onboarding = frappe.whitelist()(complete_onboarding)
    get_available_archetypes = frappe.whitelist()(get_available_archetypes)
    skip_onboarding = frappe.whitelist()(skip_onboarding)
    reset_onboarding = frappe.whitelist()(reset_onboarding)
    preview_archetype = frappe.whitelist()(preview_archetype)


# Apply whitelist decorators if frappe is available
try:
    _wrap_for_whitelist()
except ImportError:
    pass  # Decorators will be added when module is used in Frappe context

File: frontend/pim-ui/src/types/onboarding.ts
/**
 * Onboarding-related type definitions for the SaaS onboarding wizard.
 *
 * Maps to backend:
 * - PIM Onboarding State DocType (12-step state machine)
 * - Onboarding API (frappe_pim.pim.api.onboarding)
 * - Template Engine service
 */

import type { FrappeEntity } from './index'

// ============================================================================
// Onboarding Steps
// ============================================================================

/** All valid onboarding step identifiers, matching backend ONBOARDING_STEPS */
export type OnboardingStepName =
  | 'pending'
  | 'company_info'
  | 'industry_selection'
  | 'product_structure'
  | 'channel_setup'
  | 'workflow_preferences'
  | 'compliance_setup'
  | 'template_applied'
  | 'customization_review'
  | 'first_data'
  | 'guided_tour'
  | 'completed'

/** Ordered list of onboarding steps (mirrors backend constant) */
export const ONBOARDING_STEPS: readonly OnboardingStepName[] = [
  'pending',
  'company_info',
  'industry_selection',
  'product_structure',
  'channel_setup',
  'workflow_preferences',
  'compliance_setup',
  'template_applied',
  'customization_review',
  'first_data',
  'guided_tour',
  'completed',
] as const

/** Steps that collect user form data (mirrors backend STEP_DATA_FIELDS) */
export const STEP_DATA_FIELDS: Partial<Record<OnboardingStepName, string>> = {
  company_info: 'company_info_data',
  industry_selection: 'industry_selection_data',
  product_structure: 'product_structure_data',
  channel_setup: 'channel_setup_data',
  workflow_preferences: 'workflow_preferences_data',
  compliance_setup: 'compliance_setup_data',
  customization_review: 'customization_review_data',
  first_data: 'first_data_data',
  guided_tour: 'guided_tour_data',
}

// ============================================================================
// Onboarding State
// ============================================================================

/** PIM Onboarding State - DocType entity */
export interface PIMOnboardingState extends FrappeEntity {
  user: string
  current_step: OnboardingStepName
  is_completed: boolean
  is_skipped: boolean
  started_at?: string
  completed_at?: string
  selected_archetype?: string
  template_applied: boolean
  template_applied_at?: string
  template_result?: string
  progress_percent: number
  steps_completed?: string
  last_step_completed_at?: string
  error_log?: string
  /** Step data JSON fields */
  company_info_data?: string
  industry_selection_data?: string
  product_structure_data?: string
  channel_setup_data?: string
  workflow_preferences_data?: string
  compliance_setup_data?: string
  customization_review_data?: string
  first_data_data?: string
  guided_tour_data?: string
}

// ============================================================================
// API Response Types
// ============================================================================

/** Step detail in the onboarding state summary */
export interface OnboardingStepDetail {
  name: OnboardingStepName
  index: number
  is_completed: boolean
  is_current: boolean
  has_data: boolean
}

/** Onboarding state summary returned by backend APIs */
export interface OnboardingStateSummary {
  user: string
  current_step: OnboardingStepName | 'not_started'
  is_completed: boolean
  is_skipped: boolean
  progress_percent: number
  selected_archetype?: string | null
  template_applied: boolean
  started_at?: string | null
  completed_at?: string | null
  completed_steps: string[]
  next_step?: OnboardingStepName | null
  previous_step?: OnboardingStepName | null
  total_steps: number
  steps: OnboardingStepDetail[]
  /** Included when resuming a step that has saved data */
  current_step_data?: Record<string, unknown>
}

/** Step metadata from get_onboarding_steps API */
export interface OnboardingStepMetadata {
  name: OnboardingStepName
  index: number
  has_data_field: boolean
  data_field?: string | null
}

// ============================================================================
// Step Form Data Types
// ============================================================================

/** Company info step data */
export interface CompanyInfoData {
  company_name: string
  industry?: string
  company_size?: 'small' | 'medium' | 'large' | 'enterprise'
  country?: string
  currency?: string
  website?: string
}

/** Industry selection step data */
export interface IndustrySelectionData {
  archetype: string
  sub_industry?: string
  product_count_estimate?: string
  has_variants?: boolean
}

/** Product structure step data */
export interface ProductStructureData {
  use_families?: boolean
  use_categories?: boolean
  variant_levels?: number
  use_sku_generation?: boolean
  sku_pattern?: string
}

/** Channel setup step data */
export interface ChannelSetupData {
  channels?: string[]
  primary_channel?: string
  marketplace_integrations?: string[]
}

/** Workflow preferences step data */
export interface WorkflowPreferencesData {
  require_approval?: boolean
  auto_sync_to_erp?: boolean
  sync_direction?: 'PIM Master' | 'ERP Master' | 'Bidirectional'
  enable_quality_scoring?: boolean
}

/** Compliance setup step data */
export interface ComplianceSetupData {
  regulatory_standards?: string[]
  certifications_required?: string[]
  enable_safety_warnings?: boolean
}

/** Union of all step form data types */
export type StepFormData =
  | CompanyInfoData
  | IndustrySelectionData
  | ProductStructureData
  | ChannelSetupData
  | WorkflowPreferencesData
  | ComplianceSetupData
  | Record<string, unknown>

// ============================================================================
// Industry Archetype Types
// ============================================================================

/** Available archetype information */
export interface ArchetypeInfo {
  archetype: string
  label: string
  description: string
  version?: string
}

/** Response from get_available_archetypes API */
export interface AvailableArchetypesResponse {
  archetypes: ArchetypeInfo[]
  total: number
}

/** Section preview in archetype template */
export interface ArchetypeSectionPreview {
  count: number
  items: string[]
}

/** Response from preview_archetype API */
export interface ArchetypePreviewResponse {
  archetype: string
  label: string
  description: string
  version?: string
  extends?: string
  sections: Record<string, ArchetypeSectionPreview>
}

/** Template application result from apply_archetype_template API */
export interface TemplateApplicationResult {
  success: boolean
  archetype: string
  status: 'completed' | 'partial' | 'failed'
  entities_created: number
  entities_skipped: number
  entities_failed: number
  details: Record<string, TemplateEntityDetail>
  errors: string[]
  messages: string[]
  dry_run?: boolean
}

/** Per-entity-type details in template application result */
export interface TemplateEntityDetail {
  created: number
  skipped: number
  failed: number
  items: string[]
}

// ============================================================================
// UI Step Types (for the Vue.js wizard component)
// ============================================================================

/** Onboarding step definition for the UI wizard */
export interface OnboardingWizardStep {
  id: string
  title: string
  description: string
  component: string
  isCompleted: boolean
  isActive: boolean
  isFutureStep: boolean
  hasDataField?: boolean
}
